README.md
# Prando

## Overview

Prando stands for **P**atch, **R**ename, **and** **O**rganize. It's a tool for batch manipulating STFS packages (specifically XBLA, title updates, and downloadable content). It runs on Windows, Mac, and Linux.

**[Java Runtime Environment](https://www.java.com/download/) 8 or later is required for Prando to run.**

## Features

* Patch (unlock) packages for use on JTAG and RGH consoles.
* Rename packages to their proper filename (42 characters long).
* Organize packages into their correct directories so you can easily transfer them to a JTAG or RGH console.
* View multiple packages in a customizable, sortable, searchable table. Click a table header to sort the table in ascending or descending order.
* Select a row from the table to show more information about a package including all 16 License Entries in hexadecimal.

Note: Prando does not patch or rename title updates. You can safely load title updates with other Content Types.

## Usage

1. Run Prando by double clicking on Prando.jar.
2. Choose a folder to recursively look for STFS packages in or choose a single file for Prando to load.
3. View data for individual packages by selecting a row in the table. Click and drag to the right on a text field in the bottom right corner of Prando if the data is larger than the size of a text field to view the rest of it.
4. Check the checkboxes of operations you would like to perform on the packages (Patch, Rename, and/or Organize).
5. Press Start to perform the operations. If Organize has been selected, also choose a directory where you would like the newly organized packages to be placed.

## Settings

Open Prando.jar using a file archiver program and you will find a Settings.properties file in the settings folder. You can edit the contents of Settings.properties with a text editor, save it, then update Prando.jar with the newly saved file to modify Prando's settings.

### Table Column Positions

Edit these settings to add, remove, and rearrange the table column headers. Order the columns you want Prando to display numerically in ascending order and they will appear in that order from left to right. Set a column to 0 if you do not want that column to show up in the table.

(Defaults)  
Display_Name=1  
Title_Name=2  
File_Location=3  
Proper_Filename=0  
Title_ID=0  
Content_ID=0  
Media_ID=0  
Update_Number=0  
Content_Type=0  
Status=0  
File_Size=0  
STFS_Type=0  

Note: File Size in the table shows the size in bytes.

#### Status Column

* If the package has 6 bytes zeroed out in the header (from 0x6 to 0xB inclusive) the Status will be _Broken Header_. Some tools similar to Prando zero out these bytes in the header unnecessarily as a part of their unlocking process. This Status also helps identify packages that have already been unlocked by programs similar to Prando. Packages labeled with _Broken Header_ will still work fine on JTAG and RGH consoles.
* If the package does not have those 6 bytes zeroed out, the Status will be _OK_.

### General

#### Open_Path
* The _Select A File/Directory_ dialog you see when you open Prando will start in the folder you specify for this setting. The default setting is your home directory.

#### Organize_Path
* If the organize checkbox is checked and you click Start, Prando will prompt you to choose a directory. The _Organize File(s)_ dialog will start in the folder you specify for this setting. The default setting is your home directory.

#### Kilobyte_Size
* Set how many bytes are in a kilobyte for your operating system. This number can either be 1024 (for Windows and some Linux distributions) or 1000 (for Mac and some Linux distributions). This setting affects how package filesizes are calculated in Prando. (Default = 1024)

#### Search_Depth
* Set how many folders deep you want to recursively search for packages. Set it to 0 if you do not want to limit the search depth. (Default = 0)

#### Enable_Safe_Mode
* Disables the Start button so that you won't accidentally patch, rename, or organize packages. Useful if you only want to view package information using Prando. (Default = false)

#### Select_First_Item_After_Load
* If this is true, then the first item in the table will be selected after Prando finishes loading. This makes it easier to quickly scroll through the table using your keyboard's arrow keys. (Default = false)

## Screenshots

![Screenshot-1](/screenshots/Screenshot-1.png?raw=true)
![Screenshot-2](/screenshots/Screenshot-2.png?raw=true)

CHANGELOG.md
# Changelog

### v2.1.1

* Updated font file to latest version

### v2.1.0
* Increased width and height of the main window
* Set GUI colors to gray/white, removed extra color schemes
* Removed support for CON files
* Renamed some column headers to clarify what they display
* Renamed some settings to clarify what they're for
* Renamed the settings file
* Removed Fully_Organize setting, files will now be fully organized by default when organizing
* Reduced supported content types down to XBLA, title updates, and downloadable content (000D0000, 000B0000, 00000002), removed user configurable Whitelist setting
* Added more details to README
* Added this CHANGELOG
* Fixed some settings not being set properly when entries were missing from Settings.properties

### v2.0.3
* Minor color coordination change

### v2.0.2
* Minor code cleanup

### v2.0.1
* Fixed old title updates being labeled as new title updates internally which was causing them to be organized into the wrong folder structure

### v2.0.0
* Reorganized a bunch of code, made it more readable, and made the coding style more consistent
* Added Orig Filename column in table
* Added Search_Depth setting
* Added white border around thumbnails when no color scheme is set
* Removed Short Directory info field and added Title ID and Content ID info fields
* Removed button that expanded and shrunk the main window, the main window is now in expanded mode permanently
* Removed Start_Extended setting
* Removed error checking for some noncritical settings. If a setting is malformed or invalid in Settings.prando the default for that setting will be loaded instead of showing an error.
* Renamed Open_Directory and Organize_Directory settings to Open_Path and Organize_Path
* Changed font on file chooser when picking an organize location to the correct font
* Corrected filesize calculation
* File choosers now open in detailed view instead of list view

### v1.2.0
* Added ability to right click on thumbnail icons and save them as a PNG

### v1.1.0
* Added three new settings: Select_First_Item, Fully_Organize, & Start_Extended

### v1.0.0
* Initial Release

src/prando/Settings.java
package prando;

import java.awt.Font;
import java.awt.FontFormatException;
import java.io.IOException;
import java.io.InputStream;
import java.util.Arrays;
import java.util.Properties;
import java.util.Scanner;
import prando.stfs.ContentType;
import prando.util.Utilities;

public class Settings
{
    private Font font;
    private String Open_Path = System.getProperty("user.home");
    private String Organize_Path = System.getProperty("user.home");
    private Kilobyte Kilobyte_Size = Kilobyte.KB1024;
    private int Search_Depth = Integer.MAX_VALUE;
    private boolean Enable_Safe_Mode = false;
    private boolean Select_First_Item_After_Load = false;
    private byte[] License = new byte[]{-1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
        -1, -1, 0, 0, 0, 0};

    public Font getFont() { return font; }
    public String getOpenPath() { return Open_Path; }
    public String getOrganizePath() { return Organize_Path; }
    public Kilobyte getKilobyteSize() { return Kilobyte_Size; }
    public int getSearchDepth() { return Search_Depth; }
    public boolean getEnableSafeMode() { return Enable_Safe_Mode; }
    public boolean getSelectFirstItemAfterLoad() {
        return Select_First_Item_After_Load; }
    public byte[] getLicense() { return License; }

    public enum Kilobyte
    {
        KB1000(1000), KB1024(1024);

        private int byteCount;

        private Kilobyte(int bc)
        {
            byteCount = bc;
        }//constructor

        public int getByteCount()
        {
            return byteCount;
        }//getByteCount
    }//Kilobyte

    public enum Column
    {
        Display_Name(1),
        Title_Name(2),
        File_Location(3),
        Proper_Filename(0),
        Title_ID(0),
        Content_ID(0),
        Media_ID(0),
        Update_Number(0),
        Content_Type(0),
        Status(0),
        File_Size(0),
        STFS_Type(0);

        private int position;
        private int defaultPosition;

        private Column(int pos)
        {
            defaultPosition = pos;
            position = pos;
        }//constructor

        public int getPosition() { return position; }
        public void setPosition(int pos) { position = pos; }
        public void resetPosition() { position = defaultPosition; }
    }//Column

    public void readAllSettings()
    {
        readFontFromResource();
        readTableColumnPositionSettings();
        readOpenPathSetting();
        readOrganizePathSetting();
        readKilobyteSizeSetting();
        readSearchDepthSetting();
        readEnableSafeModeSetting();
        readSelectFirstItemAfterLoadSetting();
        readLicenseSetting();
    }//readAllSettings

    private void readFontFromResource()
    {
        try(InputStream is = getClass().getResourceAsStream(
                "/resources/Font.ttf"))
        {
            font = Font.createFont(Font.TRUETYPE_FONT, is).deriveFont(
                    Font.PLAIN, 13);
        }//try

        catch(FontFormatException | IOException | NullPointerException ex)
        {
            Utilities.quit(null, "Font file not found. Exiting.");
        }//catch
    }//readFontFromResource

    private void readTableColumnPositionSettings()
    {
        try(InputStream is = getClass().getResourceAsStream(
                "/settings/Settings.properties"))
        {
            Properties config = new Properties();
            config.load(is);

            for(Column c : Column.values())
            {
                int pos = Integer.parseInt(config.getProperty(c.toString()));

                if(pos < 0 || pos > 12)
                {
                    throw new NumberFormatException();
                }//if

                c.setPosition(pos);
            }//for

            checkColumnPositionValidity();
        }//try

        catch(IOException | NullPointerException | NumberFormatException ex)
        {
            resetColumns();
        }//catch
    }//readTableColumnPositionSettings

    private void checkColumnPositionValidity()
    {
        boolean noColumnsActive = true;

        for(Column c : Column.values())
        {
            if(c.getPosition() != 0)
            {
                noColumnsActive = false;
                break;
            }//if
        }//for

        if(noColumnsActive)
        {
            resetColumns();
            return;
        }//if

        boolean hasDuplicateColumn = false;

        loop:
        for(Column c1 : Column.values())
        {
            if(c1.getPosition() != 0)
            {
                for(Column c2 : Column.values())
                {
                    if(c1 != c2 && c1.getPosition() == c2.getPosition())
                    {
                        hasDuplicateColumn = true;
                        break loop;
                    }//if
                }//for
            }//if
        }//for

        if(hasDuplicateColumn)
        {
            resetColumns();
        }//if
    }//checkColumnPositionValidity

    private void resetColumns()
    {
        for(Column c : Column.values())
        {
            c.resetPosition();
        }//for
    }//resetColumns()

    private void readOpenPathSetting()
    {
        try(InputStream is = getClass().getResourceAsStream(
                "/settings/Settings.properties"))
        {
            Properties config = new Properties();
            config.load(is);
            String openPath = config.getProperty("Open_Path");

            if(openPath != null)
            {
                Open_Path = openPath;
            }//if
        }//try

        catch(IOException ex){}//catch
    }//readOpenPathSetting

    private void readOrganizePathSetting()
    {
        try(InputStream is = getClass().getResourceAsStream(
                "/settings/Settings.properties"))
        {
            Properties config = new Properties();
            config.load(is);
            String organizePath = config.getProperty("Organize_Path");

            if(organizePath != null)
            {
                Organize_Path = organizePath;
            }//if
        }//try

        catch(IOException ex){}//catch
    }//readOrganizePathSetting

    private void readKilobyteSizeSetting()
    {
        try(InputStream is = getClass().getResourceAsStream(
                "/settings/Settings.properties"))
        {
            Properties config = new Properties();
            config.load(is);
            int kilobyteSize = Integer.parseInt(config.getProperty(
                    "Kilobyte_Size"));

            for(Kilobyte kilobyte : Kilobyte.values())
            {
                if(kilobyteSize == kilobyte.getByteCount())
                {
                    Kilobyte_Size = kilobyte;
                    break;
                }//if
            }//for
        }//try

        catch(IOException | NullPointerException | NumberFormatException ex){}
    }//readKilobyteSizeSetting

    private void readSearchDepthSetting()
    {
        try(InputStream is = getClass().getResourceAsStream(
                "/settings/Settings.properties"))
        {
            Properties config = new Properties();
            config.load(is);
            int searchDepth = Integer.parseInt(config.getProperty(
                    "Search_Depth"));

            if(searchDepth > 0)
            {
                Search_Depth = searchDepth;
            }//if
        }//try

        catch(IOException | NullPointerException | NumberFormatException ex){}
    }//readSearchDepthSetting

    private void readEnableSafeModeSetting()
    {
        try(InputStream is = getClass().getResourceAsStream(
                "/settings/Settings.properties"))
        {
            Properties config = new Properties();
            config.load(is);
            Enable_Safe_Mode = Boolean.parseBoolean(config.getProperty(
                    "Enable_Safe_Mode"));
        }//try

        catch(IOException | NullPointerException ex){}//catch
    }//readEnableSafeModeSetting

    private void readSelectFirstItemAfterLoadSetting()
    {
        try(InputStream is = getClass().getResourceAsStream(
                "/settings/Settings.properties"))
        {
            Properties config = new Properties();
            config.load(is);
            Select_First_Item_After_Load = Boolean.parseBoolean(
                    config.getProperty("Select_First_Item_After_Load"));
        }//try

        catch(IOException | NullPointerException ex){}//catch
    }//readSelectFirstItemAfterLoadSetting

    private void readLicenseSetting()
    {
        try(InputStream is = getClass().getResourceAsStream(
                "/settings/Settings.properties"))
        {
            Properties config = new Properties();
            config.load(is);
            String bytes = config.getProperty("License");

            if(bytes != null)
            {
                Scanner scan = new Scanner(bytes);
                scan.useDelimiter(",");

                byte[] license = new byte[16];

                for(int i = 0; i < license.length; i++)
                {
                    license[i] = Byte.parseByte(scan.next());
                }//for

                License = license;
            }//if
        }//try

        catch(IOException | NullPointerException ex){}//catch
    }//readLicenseSetting

    public boolean isWhitelisted(String path)
    {
        byte[] b = new byte[4];
        Utilities.readHex(path, b, 0x344, 4);

        for(ContentType contentType : ContentType.values())
        {
            if(Arrays.equals(b, contentType.getByteRepresentation()))
            {
                return true;
            }//if
        }//for

        return false;
    }//isWhitelisted
}//Settings